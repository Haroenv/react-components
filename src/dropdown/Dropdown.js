import React from 'react';
import { props, t, skinnable } from '../utils';
import Select from 'react-select';
import find from 'lodash/find';
import cx from 'classnames';
import { warn } from '../utils/log';

const PropTypes = {
  value: t.maybe(t.union([t.Number, t.String, t.Object])),
  onChange: t.maybe(t.Function),
  options: t.Array,
  id: t.maybe(t.String),
  className: t.maybe(t.String),
  style: t.maybe(t.Object)
};

@skinnable()
@props(PropTypes, { strict: false })
export default class Dropdown extends React.Component {

  componentDidMount() {
    this.logWarnings();
  }

  logWarnings = () => {
    if (this.props.children) {
      warn('You\'re passing children. Not expected behaviour');
    }
  };

  getValue = () => (
    this.props.valueLink ? this.props.valueLink.value : this.props.value
  );

  valueToOption = (value, options) => {
    if (t.String.is(value) || t.Number.is(value)) {
      return find(options, { value });
    }
    return value;
  };

  _onChange = () => this.props.valueLink ? this.props.valueLink.requestChange : this.props.onChange;

  _onBlur = () => this.forceUpdate();

  getLocals() {
    const {
      className,
      options,
      valueLink,
      ...props
    } = this.props;

    return {
      ...props,
      options,
      className: cx('dropdown', className),
      value: this.valueToOption(this.getValue(), options),
      onChange: this._onChange(),
      onBlur: this._onBlur
    };
  }

  template(locals) {
    return <Select {...locals} />;
  }
}
