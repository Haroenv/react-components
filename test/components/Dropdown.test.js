import React from 'react';
import render from '../render';

import Dropdown from '../../src/dropdown';

const exampleProps = {
  id: '12345',
  className: 'fancy-class-name',
  style: { margin: 10, position: 'relative' },
  value: { value: 'test', label: 'Test', foo: 'bar' },
  fakeProp: true,
  options: [
    { value: 'test', label: 'Test' },
    { value: 'test1', label: 'Test1' },
    { value: 'test2', label: 'Test2' }
  ]
};

const dropdown = new Dropdown(exampleProps);

describe('Dropdown', () => {

  it('renders correctly', () => {
    const tree = render(
      <Dropdown {...exampleProps} />
    );
    expect(tree).toMatchSnapshot();
  });

  describe('getLocals', () => {

    it('computes className', () => {
      const { className } = dropdown.getLocals();
      expect(className).toContain('dropdown');
      expect(className).toContain('fancy-class-name');
    });

    it('doesn\'t alter object value', () => {
      const { value } = dropdown.getLocals();
      expect(typeof value).toBe('object');
      expect(value).toEqual(exampleProps.value);
    });

    it('computes value from a string', () => {
      const dropdown = new Dropdown({
        value: 'test',
        options: [
          { value: 'test', label: 'Test' },
          { value: 'test1', label: 'Test1' },
          { value: 'test2', label: 'Test2' }
        ]
      });
      const { value, options } = dropdown.getLocals();
      expect(typeof value).toBe('object');
      expect(value).toEqual(options[0]);
    });

    it('computes value from a number', () => {
      const dropdown = new Dropdown({
        value: 2,
        options: [
          { value: 0, label: 'Test' },
          { value: 1, label: 'Test1' },
          { value: 2, label: 'Test2' }
        ]
      });
      const { value, options } = dropdown.getLocals();
      expect(typeof value).toBe('object');
      expect(value).toEqual(options[2]);
    });
  });

});
