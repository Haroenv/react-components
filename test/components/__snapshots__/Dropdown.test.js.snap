exports[`Dropdown renders correctly 1`] = `
<div
  className="Select dropdown fancy-class-name is-medium Select--single has-value"
  style={undefined}>
  <div
    className="Select-control"
    onKeyDown={[Function bound handleKeyDown]}
    onMouseDown={[Function bound handleMouseDown]}
    onTouchEnd={[Function bound handleTouchEnd]}
    onTouchMove={[Function bound handleTouchMove]}
    onTouchStart={[Function bound handleTouchStart]}
    style={
      Object {
        "margin": 10,
        "position": "relative"
      }
    }>
    <span
      className="Select-multi-value-wrapper"
      id="react-select-2--value">
      <div
        className="Select-value"
        style={undefined}
        title={undefined}>
        <span
          aria-selected="true"
          className="Select-value-label"
          id="react-select-2--value-item"
          role="option">
          Test
        </span>
      </div>
      <div
        aria-activedescendant="react-select-2--value"
        aria-expanded={false}
        aria-owns="react-select-2--value"
        aria-readonly="false"
        className="Select-input"
        onBlur={[Function bound handleInputBlur]}
        onFocus={[Function bound handleInputFocus]}
        role="combobox"
        style={
          Object {
            "border": 0,
            "display": "inline-block",
            "width": 1
          }
        }
        tabIndex={0} />
    </span>
    <span
      className="Select-arrow-zone"
      onMouseDown={[Function bound handleMouseDownOnArrow]}>
      <span
        className="Select-arrow"
        onMouseDown={[Function bound handleMouseDownOnArrow]} />
    </span>
  </div>
</div>
`;
